### PR TITLE
Fixes #429 by adding -fPIC to src/Makevars.in

### LINKED ISSUE #429: Linux build failure as of 7a0c518
The commit 7a0c518 breaks my linux build, and I can no longer install Nimble. This is a rather large commit, and it's difficult to pinpoint the error.
```
$ git checkout 7a0c518  # A.k.a. devel at the momement.
$ R CMD build
...
$ R CMD INSTALL nimble_0.6-6.tar.gz
* installing to library ‘/home/fritz/R/x86_64-pc-linux-gnu-library/3.4’
* installing *source* package ‘nimble’ ...
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
working on eigen
checking how to run the C++ preprocessor... clang++ -Qunused-arguments -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking Eigen/Dense usability... no
checking Eigen/Dense presence... no
checking for Eigen/Dense... no
Using the version of Eigen provided with NIMBLE
Using CppAD directory 
NIMBLE_DIR = /home/fritz/R/x86_64-pc-linux-gnu-library/3.4/nimble
ENABLE_LIB=true
configure: creating ./config.status
config.status: creating inst/make/Makevars
config.status: creating inst/make/Makevars_lib
config.status: creating R/config.R
config.status: creating src/Makevars
Building libnimble
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c RcppUtils.cpp -o RcppUtils.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c EigenTypedefs.cpp -o EigenTypedefs.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c eigenUsingClasses.cpp -o eigenUsingClasses.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c RcppNimbleUtils.cpp -o RcppNimbleUtils.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c Utils.cpp -o Utils.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c NamedObjects.cpp -o NamedObjects.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c ModelClassUtils.cpp -o ModelClassUtils.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c accessorClasses.cpp -o accessorClasses.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c dists.cpp -o dists.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c nimDists.cpp -o nimDists.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c dllFinalizer.cpp -o dllFinalizer.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c nimbleGraph.cpp -o nimbleGraph.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c smartPtrs.cpp -o smartPtrs.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c predefinedNimbleLists.cpp -o predefinedNimbleLists.o
g++  -I/usr/share/R/include -DNDEBUG -I"../include"     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c nimOptim.cpp -o nimOptim.o
linking libnimble.so
Linux   
g++  -shared -o libnimble.so  RcppUtils.o EigenTypedefs.o eigenUsingClasses.o RcppNimbleUtils.o Utils.o NamedObjects.o ModelClassUtils.o accessorClasses.o dists.o nimDists.o dllFinalizer.o nimbleGraph.o smartPtrs.o predefinedNimbleLists.o nimOptim.o 
rm *.o
** libs
clang++ -Qunused-arguments -I/usr/share/R/include -DNDEBUG -I../inst/include  -I     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c nimble.cpp -o nimble.o
clang++ -Qunused-arguments -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o nimble.so nimble.o -L../inst/CppCode -lnimble -Wl,-rpath=/home/fritz/R/x86_64-pc-linux-gnu-library/3.4/nimble/CppCode -llapack -lblas -L/usr/lib/R/lib -lR
/usr/bin/ld: nimble.o: relocation R_X86_64_32 against `CallEntries' can not be used when making a shared object; recompile with -fPIC
nimble.o: error adding symbols: Bad value
clang: error: linker command failed with exit code 1 (use -v to see invocation)
/usr/share/R/share/make/shlib.mk:6: recipe for target 'nimble.so' failed
make: *** [nimble.so] Error 1
make: Target 'all' not remade because of errors.
ERROR: compilation failed for package ‘nimble’
* removing ‘/home/fritz/R/x86_64-pc-linux-gnu-library/3.4/nimble’
* restoring previous ‘/home/fritz/R/x86_64-pc-linux-gnu-library/3.4/nimble’
```

### PR DESCRIPTION
**Pull request by fritzo**
Tested on Ubuntu 16.04
```
$ R --version ; g++ --version ; uname -a
R version 3.4.0 (2017-04-21) -- "You Stupid Darkness"
Copyright (C) 2017 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under the terms of the
GNU General Public License versions 2 or 3.
For more information about these matters see
http://www.gnu.org/licenses/.

g++ (Ubuntu 5.4.1-2ubuntu1~16.04) 5.4.1 20160904
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Linux nimbook 4.8.0-52-generic #55~16.04.1-Ubuntu SMP Fri Apr 28 14:36:29 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

### CODE REVIEW

**Comment by perrydv** [comment]
Should we be conservative and only insert `-fPIC` on linux?  We could do so by modifying the `configure.ac` and adding a new replacement field like @FPIC@.  It may not be necessary, but it might be safer to modify only the case we know to be an issue.  What do you think?

**Comment by paciorek** [comment]
Some concerns here. I have not followed the details and have forgotten what little I've ever known about -fpic so this just serves as a report of what I am seeing:

1) R CMD check is not happy with Makevars.in
```
* checking compilation flags in Makevars ... WARNING
Non-portable flags in variable 'PKG_CPPFLAGS':
  -fPIC
```
I think we need to see what the R extensions manual says about pic and fear that they are not going to like the -fPIC...

2) compilation for libnimble is still using -fpic, while compilation for nimble.so has both pic and PIC: "-fpic -fPIC".

**Comment by perrydv** [comment]
If it is non-portable, another solution would be to actually consolidate some of our `extern`ed functions.  But that would be a twist to do before the next release.

**Comment by paciorek** [comment]
Given the portability flag, should we go with @perrydv's suggestion to use @FPIC@ via configure.ac? 

Or, this might be one to call in our lifeline to Duncan as well. 

@fritzo @perrydv: thoughts? Would be nice to resolve this one way or another so we can build from devel on Linux.

**Comment by fritzo** [comment]
@paciorek Perry's suggestion sounds reasonable. I can get to it early next week, or feel free to do so over the weekend.

**Comment by paciorek** [comment]
I'm going to leave it to you. Thanks!

Do you see a way to avoid having "-fPIC -fpic" in the nimble.o compilation step during install?

**Comment by fritzo** [comment]
I don't see a way, but I'm really bumping my way around blindly here :-/

**Comment by paciorek** [comment]
ok, I think I'll put in an email to Duncan to see if he has any thoughts
here, as otherwise I'd just spend time also bumping around blindly too.

**Comment by fritzo** [comment]
Well I've added a `case` statement in our `configure.ac` as @duncantl suggested and used it in our Makevars.in as @perrydv suggested. Any thoughts?

File: packages/nimble/configure.ac [hunk 1]
```diff
@@ -12,6 +12,20 @@ AC_PROG_CXX
 
 NEED_MAKEVARS_FILE=TRUE
```
Comment (fritzo) [inline packages/nimble/configure.ac]: This is the new section I've added as per @duncantl 's suggestion.

**Comment by paciorek** [comment]
ok, R CMD check seems ok with the changes. 

@duncantl do you see any way around (or are you unconcerned) with having both pic and PIC flags in our compilation of nimble.cpp? This is the compilation step on my Linux box when installing the package:
```
g++  -I/usr/share/R/include -DNDEBUG -I../inst/include  -I -fPIC     -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c nimble.cpp -o nimble.o
```

**Comment by fritzo** [comment]
@paciorek May I merge this, or would you prefer to hold off?

**Comment by paciorek** [comment]
Let's merge. I wasn't sure from the issue/pr comments if you had actually
talked to Duncan or not.

In any event, it would be good not to have devel broken for Linux. We can
iron out dangling issues or revisions based on Duncan suggestions as part
of another PR.

**Pull request merged**